### PR TITLE
op-node: Refactor types and L1/L2 Source

### PR DESCRIPTION
**Pull request by trianglesphere**
**Description**
This PR modifies moves types into the `eth` package from the `derive` package,
updates some handling of the execution engine API error results, and adds some
testutils.

**Additional context**
These changes are pulled from the full batch derivation PR.

### CODE REVIEW

File: op-node/l2/source.go [hunk 1]
```diff
@@ -88,24 +89,12 @@ func (s *Source) ForkchoiceUpdate(ctx context.Context, fc *eth.ForkchoiceState,
...
-	case eth.ExecutionAccepted, eth.ExecutionInvalidTerminalBlock, eth.ExecutionInvalidBlockHash:
-		// ACCEPTED, INVALID_TERMINAL_BLOCK, INVALID_BLOCK_HASH are only for execution
-		return nil, fmt.Errorf("unexpected %s status, could not update forkchoice: %v", result.PayloadStatus.Status, err)
-	case eth.ExecutionInvalid:
-		return nil, fmt.Errorf("cannot update forkchoice, block is invalid: %v", err)
-	case eth.ExecutionValid:
-		return &result, nil
-	default:
-		return nil, fmt.Errorf("unknown forkchoice status on %s: %q, ", fc.SafeBlockHash, string(result.PayloadStatus.Status))
+		return nil, err
```
Comment (protolambda) [inline op-node/l2/source.go:92]: This does change the error behavior. You should call `eth.ForkchoiceUpdateErr` to transform the status into an error to preserve the current behavior. In the `batch-derivation` branch it changes again though, the difference between an invalid payload and a RPC connection error shouldn't be hidden from the derivation code.

File: op-node/l2/source.go [hunk 2]
```diff
@@ -116,25 +105,9 @@ func (s *Source) NewPayload(ctx context.Context, payload *eth.ExecutionPayload)
...
-		return fmt.Errorf("execution payload %s has INVALID BLOCKHASH! %v", payload.BlockHash, result.ValidationError)
-	case eth.ExecutionInvalidTerminalBlock:
-		return fmt.Errorf("engine is misconfigured. Received invalid-terminal-block error while engine API should be active at genesis. err: %v", result.ValidationError)
-	case eth.ExecutionAccepted:
-		return fmt.Errorf("execution payload cannot be validated yet, latest valid hash is %s", result.LatestValidHash)
-	default:
-		return fmt.Errorf("unknown execution status on %s: %q, ", payload.ID(), string(result.Status))
+		return nil, fmt.Errorf("failed to execute payload: %v", err)
 	}
+	return &result, nil
```
Comment (protolambda) [inline op-node/l2/source.go:110]: Same here, but `eth.NewPayloadErr`

File: op-node/l2/source.go (hunk 1 above)
Comment (trianglesphere) [reply op-node/l2/source.go:92]: done

File: op-node/rollup/driver/step.go [hunk 3]
```diff
@@ -389,10 +390,12 @@ func (d *outputImpl) insertHeadBlock(ctx context.Context, fc eth.ForkchoiceState
 		d.log.Error("Dropped deposits when executing L2 block")
 	}
 
-	err = d.l2.NewPayload(ctx, payload)
-	if err != nil {
+	if status, err := d.l2.NewPayload(ctx, payload); err != nil {
+		return nil, fmt.Errorf("failed to insert execution payload: %w", err)
+	} else if err := eth.ForkchoiceUpdateErr(*status); err != nil {
```
Comment (protolambda) [inline op-node/rollup/driver/step.go:395]:
```suggestion
	} else if err := eth.NewPayloadErr(payload, status); err != nil {
```

File: op-node/rollup/driver/step.go [hunk 4]
```diff
@@ -55,8 +54,10 @@ func lastDeposit(txns []eth.Data) (int, error) {
 
 func (d *outputImpl) processBlock(ctx context.Context, l2Head eth.L2BlockRef, l2SafeHead eth.BlockID, l2Finalized eth.BlockID, payload *eth.ExecutionPayload) error {
 	d.log.Info("processing new block", "parent", payload.ParentID(), "l2Head", l2Head, "id", payload.ID())
-	if err := d.l2.NewPayload(ctx, payload); err != nil {
-		return fmt.Errorf("failed to insert new payload: %v", err)
+	if status, err := d.l2.NewPayload(ctx, payload); err != nil {
+		return fmt.Errorf("failed to insert new payload: %w", err)
+	} else if err := eth.ForkchoiceUpdateErr(*status); err != nil {
```
Comment (trianglesphere) [inline op-node/rollup/driver/step.go:59]:
```suggestion
	} else if err := eth.NewPayloadErr(payload, status); err != nil {
```

File: op-node/rollup/driver/step.go (hunk 3 above)
Comment (trianglesphere) [reply op-node/rollup/driver/step.go]: updated as well in the other place that does this

**Pull request merged**